### PR TITLE
Register `~const` preds for `Deref` adjustments in HIR typeck

### PR DESCRIPTION
**Pull request by compiler-errors**
This doesn't *do* anything yet, since `Deref` and `DerefMut` aren't constified, and we explicitly don't error on calling non-const trait methods in HIR yet -- presumably that will wait until std is re-constified. But I'm confident this logic is correct, and this (afaict?) is the only major hole left in enforcing `~const` in HIR typeck.

r? fee1-dead

### CODE REVIEW

**Comment by fee1-dead** [comment]
Haven't looked at the code closely, but this could probably use a minicore test if possible.

**Comment by compiler-errors** [comment]
I fixed up the existing minicore, moved it to an auxiliary dependency, so we can start adding real tests based off of that minicore that exercise its behavior.

File: compiler/rustc_hir_typeck/src/place_op.rs [hunk 1]
```diff
@@ -289,13 +289,10 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
-                        let ty::Ref(_, _, mutbl) = *method.sig.output().kind() else {
-                            span_bug!(
-                                self.tcx.def_span(method.def_id),
-                                "expected DerefMut to return a &mut"
-                            );
-                        };
-                        *deref = OverloadedDeref { mutbl, span: deref.span };
+                        if let ty::Ref(_, _, mutbl) = *method.sig.output().kind() {
+                            *deref = OverloadedDeref { mutbl, span: deref.span };
+                        }
```
Comment (fee1-dead) [inline compiler/rustc_hir_typeck/src/place_op.rs:294]: Why was this span_bug removed?

File: compiler/rustc_middle/src/ty/adjustment.rs [hunk 2]
```diff
@@ -124,18 +125,16 @@ pub struct OverloadedDeref {
 
 impl OverloadedDeref {
     /// Get the zst function item type for this method call.
```
Comment (fee1-dead) [inline compiler/rustc_middle/src/ty/adjustment.rs:127]: This doc should be updated

**Review by fee1-dead — CHANGES_REQUESTED** [review]

File: compiler/rustc_hir_typeck/src/place_op.rs (hunk 1 above)
Comment (compiler-errors) [reply compiler/rustc_hir_typeck/src/place_op.rs:294]: It's kinda pointless. I think I mixed up a bit of the refactoring from #132301, but all it's doing is asserting something we've explicitly decided not to care about (which is detecting invalid libcore items, https://github.com/rust-lang/compiler-team/issues/620).

Comment (compiler-errors) [reply compiler/rustc_hir_typeck/src/place_op.rs:294]: But whatever, I'll revert it.

**Review by fee1-dead — APPROVED** [review]

**Comment by workingjubilee** [comment]
failed in https://github.com/rust-lang/rust/pull/132551#issuecomment-2453336111

```

warning: function `const_eval_select` is never used
##[warning]  --> /Users/runner/work/rust/rust/tests/ui/traits/const-traits/effects/auxiliary/minicore.rs:465:10
   |
   |
LL | const fn const_eval_select<ARG: Tuple, F, G, RET>(

error: linking with `cc` failed: exit status: 1
   |
   |
   = note: env -u IPHONEOS_DEPLOYMENT_TARGET -u TVOS_DEPLOYMENT_TARGET -u XROS_DEPLOYMENT_TARGET LC_ALL="C" PATH="/Users/runner/work/rust/rust/build/aarch64-apple-darwin/stage2/lib/rustlib/aarch64-apple-darwin/bin:/opt/homebrew/lib/ruby/gems/3.0.0/bin:/opt/homebrew/opt/ruby@3.0/bin:/Users/runner/.local/bin:/opt/homebrew/bin:/opt/homebrew/sbin:/Users/runner/.cargo/bin:/usr/local/opt/curl/bin:/usr/local/bin:/usr/local/sbin:/Users/runner/bin:/Users/runner/.yarn/bin:/Users/runner/Library/Android/sdk/tools:/Users/runner/Library/Android/sdk/platform-tools:/Library/Frameworks/Python.framework/Versions/Current/bin:/Library/Frameworks/Mono.framework/Versions/Current/Commands:/usr/bin:/bin:/usr/sbin:/sbin:/Users/runner/.dotnet/tools" VSLANG="1033" ZERO_AR_DATE="1" "cc" "-Wl,-exported_symbols_list" "-Wl,/var/folders/g6/rgtlsw6n123b0gt5483s5_cm0000gn/T/rustclDu9Z6/list" "/var/folders/g6/rgtlsw6n123b0gt5483s5_cm0000gn/T/rustclDu9Z6/symbols.o" "/Users/runner/work/rust/rust/build/aarch64-apple-darwin/test/ui/traits/const-traits/effects/minicore-works/auxiliary/minicore.minicore.356631b40ca0b657-cgu.0.rcgu.o" "/Users/runner/work/rust/rust/build/aarch64-apple-darwin/test/ui/traits/const-traits/effects/minicore-works/auxiliary/minicore.0lca3mr32tz9pvmmfm5apoe3h.rcgu.rmeta" "-arch" "arm64" "-mmacosx-version-min=11.0.0" "-L" "/Users/runner/work/rust/rust/build/aarch64-apple-darwin/native/rust-test-helpers" "-L" "/Users/runner/work/rust/rust/build/aarch64-apple-darwin/test/ui/traits/const-traits/effects/minicore-works/auxiliary" "-o" "/Users/runner/work/rust/rust/build/aarch64-apple-darwin/test/ui/traits/const-traits/effects/minicore-works/auxiliary/libminicore.dylib" "-Wl,-dead_strip" "-dynamiclib" "-Wl,-install_name" "-Wl,@rpath/libminicore.dylib" "-nodefaultlibs"
   = note: Undefined symbols for architecture arm64:
             "dyld_stub_binder", referenced from:
                 <initial-undefines>
           ld: symbol(s) not found for architecture arm64
           clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

@bors r-

Uh can we get a ~~doctor~~ orchard farmer in the house?

**Comment by compiler-errors** [comment]
cool, looks like i was just missing a `no-prefer-dynamic` on the minicore.

**Comment by compiler-errors** [comment]
Rebased due to #132479.

@bors r=fee1-dead

**Comment by compiler-errors** [comment]
:( looks like there's still a conflict

**Comment by compiler-errors** [comment]
re-rebased

@bors r=fee1-dead

**Pull request merged**